### DISCUSSION TITLE
Timeout when connecting to IPv6 address via FTPS

### DISCUSSION THREAD

**Discussion by ribtoks** [curl/curl#15710] [Q&A]
Hi

I'm trying to help a user of my app that has connection problems when using FTPS and libcurl. Source code is basically modified version of [FTP upload](https://curl.se/libcurl/c/ftpupload.html) but with enabled FTPS, so I don't post it here for brevity. Here are the logs though (again, using tracing functions from examples):

```
11:54:42.721 info T#4956457 Connectivity::fillFtpCurlOptions - pIi[ "ftp://ftps.shutterstock.com/" ]Iip
11:54:42.721 debug T#4956457 Connectivity::fillFtpCurlOptions - EPSV mode disabled
11:54:42.721 debug T#4956457 Connectivity::fillFtpCurlOptions - Forcing FTPS
11:54:42.721 debug T#4956457 Connectivity::fillFtpCurlOptions - Verify certificate: false host: true
11:54:42.722 debug T#4956457 Connectivity::isContextValid - About to check credentials for "ftps.shutterstock.com"
11:54:42.899 info T#4956457 Connectivity::my_trace - "== Info: Host ftps.shutterstock.com:21 was resolved."
11:54:42.899 info T#4956457 Connectivity::my_trace - "== Info: IPv6: 64:ff9b::36a5:c8c3, 64:ff9b::22ef:399b, 64:ff9b::22ce:cb95"
11:54:42.899 info T#4956457 Connectivity::my_trace - "== Info: IPv4: 34.206.203.149, 54.165.200.195, 34.239.57.155"
11:54:42.899 info T#4956457 Connectivity::my_trace - "== Info:   Trying [64:ff9b::36a5:c8c3]:21..."
11:54:42.927 info T#4956457 Connectivity::my_trace - "== Info: Connected to ftps.shutterstock.com (64:ff9b::36a5:c8c3) port 21"
11:54:43.159 debug T#4956457 Connectivity::dump - <= Recv header, 0000000043 bytes (0x0000002b) 220-Welcome to t he Shutterstock  FTP site...
.... <----- here some FTP welcome message stripped
11:54:43.192 debug T#4956457 Connectivity::dump - <= Recv header, 0000000027 bytes (0x0000001b) 220 Shutterstock  FTP Site..
11:54:43.193 debug T#4956457 Connectivity::dump - => Send header, 0000000010 bytes (0x0000000a) AUTH SSL..
11:54:43.324 debug T#4956457 Connectivity::dump - <= Recv header, 0000000025 bytes (0x00000019) 234 AUTH SSL suc cessful..
11:59:42.726 info T#4956457 Connectivity::my_trace - "== Info: SSL connection timeout"
11:59:42.726 info T#4956457 Connectivity::my_trace - "== Info: closing connection #0"
11:59:42.727 info T#4956457 Connectivity::isContextValid - Credentials checking error: Timeout was reached
11:59:42.727 info T#4956457 Connectivity::isContextValid - Response code: 234
11:59:42.727 info T#4956457 Connectivity::isContextValid - Credentials checking finished "ftps.shutterstock.com"
```

SSL connection timeout happens 5 minutes later, accordingly to the logs.

My basic questions are:
- how to even debug this?
- if it's possible to make libcurl to select IPv4 address over IPv6 address? (I assume those are IP addresses returned by PASV response)
- are there any SSL curl options I should consider?

I just want to make it clear that it's likely not a curl bug, I'm just asking for help in how to debug/solve this. User reported successful connection when switching to another network (mobile hotspot), this error is happening only when using WiFi.

Thank you!

**Comment by icing** [comment]
Hey @ribtoks, I assume your log with `my_trace` is the output via `CURLOPT_DEBUGFUNCTION`? In a recent libcurl, you can raise the level of trace output with `curl_global_trace("all");` right after the global init. Maybe that helps.

For restricting the IP address family. there is `CURLOPT_IPRESOLVE` to configure that for a transfer. Look at the manpage to see examples.

**Reply by ribtoks** [reply]
Hi @icing and thanks for the fast response.

`my_trace` is the debug function, that's right. Thanks for the tip with `curl_global_trace()`, I didn't know about it's existance.

I saw `CURLOPT_IPRESOLVE` option, but I thought that maybe somehow FTP protocol was involved (e.g. PASV response, which was affected by `CURLOPT_FTP_SKIP_PASV_IP`).

Few "stupid" questions:
- Is there any way to make curl to try "next resolved IP address" if there are few and current connection failed?
- Is there any way to "prioritize" IPv4 addresses? `CURLOPT_IPRESOLVE` will enable me to completely disable IPv6, if I understand correctly, if I set it to `CURL_IPRESOLVE_V4 `

**Reply by icing** [reply]
curl does happy eyeballing, meaning it tries all resolved addresses, prioritising ipv6 and some milliseconds later in parallel ipv4 addresses. The first address that connects ends this process. If there is an error later during the protocol, this does not trigger a retry on another address.

If you want to test with a specific address, there is `CURLOPT_RESOLVE` where you can specify a list of ip addresses for a host name.

From the log, I see that the connection stalls before any data connection is attempted - at leat it looks like that. Maybe you can produce a more verbose log with higher tracing to see where the problem is.

**Reply by ribtoks** [reply]
Thank you for the suggestions. I tried to setup `CURLOPT_RESOLVE` and, judging from the logs, it seems that connection indeed cannot be established. I'm wondering if on my user's side they got a bunch of stale DNS records... Not sure there's anything I can do in this situation. Thank you anyways for your help!

**Reply by ribtoks** [reply]
Hey @icing
Since we're on this topic, I have one more strange question. Is there any way to ignore an IP address from resolve mechanism? I understand I can add override via `CURLOPT_RESOLVE`, but is there a reverse thing, meaning a list of addresses that are "ignored" from resolving?
The reason I ask is because for some reason (e.g. cached DNS records), the address is resolving to an IP address which is accessible, but malfunctioning ("business logic"-wise) and rather than timing out, I'd prefer to just ignore it. However, I don't want to "prefetch" the list of all IP addresses by manually resolving the address and then "overriding" the curl list using `_RESOLVE` option as it's more hassle.
Thank you!

**Reply by jay** [reply]
> Is there any way to ignore an IP address from resolve mechanism?

You can use [CURLOPT_OPENSOCKETFUNCTION](https://curl.se/libcurl/c/CURLOPT_OPENSOCKETFUNCTION.html) to block IP addresses. A quick and dirty way is use inet_ntop to get it in presentation format and compare. Like say the problem addresses are 70.x.x.x you could strncmp to "70.". For a correct way see #15748. You may also want to look into [CURLOPT_DNS_SHUFFLE_ADDRESSES](https://curl.se/libcurl/c/CURLOPT_DNS_SHUFFLE_ADDRESSES.html).

**Reply by ribtoks** [reply]
Hi @jay Thank you so much for taking time to dig into this. Appreciate your example as well!
